File: 卡码/代码随想录/1/07-回溯/Java/backtracking/src/main/java/com/kurumi/backtracking/_03_子集/LeetCode_0078.java
//78. 子集
package com.kurumi.backtracking._03_子集;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class LeetCode_0078 {

}
class Solution {
  Deque<Integer> path=new LinkedList<>();
  List<List<Integer>> res=new ArrayList<>();
  public void backtracking(int[] nums,int index){
    res.add(new ArrayList<>(path));
    if(index>=nums.length) return;
    for(int i=index;i<nums.length;i++){
      path.add(nums[i]);
      backtracking(nums, i+1);
      path.removeLast();
    }
  }
  public List<List<Integer>> subsets(int[] nums) {
    backtracking(nums, 0);
    return res;
  }
}

File: 卡码/代码随想录/1/07-回溯/Js/03-子集/LeetCode_0078.js
//78. 子集
const subsets = function(nums) {
  const res=new Array();
  const path=new Array();
  const backtracking=function(nums,index){
    res.push(Array.from(path));
    if(index>=nums.length) return;
    for(let i=index;i<nums.length;i++){
      path.push(nums[i]);
      backtracking(nums,i+1);
      path.pop();
    }
  }
  backtracking(nums,0);
  return res;
};